src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    /// :raises ValueError: ``to_replace`` is empty or ``node``
+    ///     has existing connections.
+    #[pyo3(
+        text_signature = "(self, to_replace, node, /, force_check_cycle=None)"
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
+
+            // DFS. If we encounter any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            while let Some(decendant) = visit_next.pop_back() {
+                if visited.is_visited(&decendant) {
+                    continue;
+                }
+
+                for edge in graph.edges(decendant) {
+                    if nodes.contains(&edge.target()) {
+                        // we found a path back to `nodes`
+                        return false;
+                    }
+
+                    visit_next.push_back(edge.target());
+                }
+
+                visited.visit(decendant);
+            }
+
+            true
+        }
+
+        macro_rules! assert_node_exists {
+            ($node_index:ident) => {
+                if self.graph.node_weight($node_index).is_none() {
+                    return Err(PyIndexError::new_err(format!(
+                        "Specified node {} is not in this graph",
+                        $node_index.index()
+                    )));
+                }
+            };
+        }
+
+        if to_replace.is_empty() {
+            return Err(PyValueError::new_err("to_replace must not be empty"));
+        }
+
+        let node_index = NodeIndex::new(node);
+        assert_node_exists!(node_index);
+
+        if self.graph.neighbors_undirected(node_index).next().is_some() {
+            return Err(PyValueError::new_err(
+                "replacement node must not have connections",
+            ));
+        }
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            HashSet::with_capacity(to_replace.capacity());
+
+        for node in to_replace {
+            let index = NodeIndex::new(node);
+            assert_node_exists!(index);
+            indices_to_remove.insert(index);
+        }
```

Comment:
I don't feel it's necessary to raise an error here, we just ignore it if an index does not belong to the graph.
```suggestion
        let indices_to_remove: HashSet<NodeIndex> = to_replace
            .into_iter()
            .filter_map(|node| {
                let index = NodeIndex::new(node);
                if self.graph.contains_node(index) {
                    Some(index)
                } else {
                    None
                }
            })
            .collect();
```

Reply:
Would you also recommend then skipping check of empty `to_replace`? We could have that as a no-op, and just note the behavior in the docstring.

Reply:
Yeah but this might actually depend on how we address this comment https://github.com/Qiskit/retworkx/pull/487#discussion_r750309097, e.g if this function is expected to add a new node, it feels better to keep the empty-check

Reply:
Hmm. I guess if `to_replace` is empty, this function behaves like `add_node`, assuming we keep the changes introduced by c8699ae. This seems reasonable to me, and allows us to skip checks. If we keep the empty check, then we also probably need to check that at least one node in `to_replace` exists, otherwise I'd think the method should also fail.

Reply:
Added behavior note in f734242. Happy to revise if you'd prefer this to work differently.